### PR TITLE
GH-260, PTRENG-3557. Password is required, validation is offloaded to the Artifactory side

### LINKED ISSUE #260: `artifactory_user` password documented as optional but is required
**Describe the bug**
When I create a user with no password, I get:

```
│ Error: no password supplied. Please use any of the terraform random password generators
│ 
│   with module.some_human.artifactory_user.user[0],
│   on modules/user/artifactory.tf line 1, in resource "artifactory_user" "user":
│    1: resource "artifactory_user" "user" {
│ 
╵
```

This is contrary to the documentation, which [states](https://registry.terraform.io/providers/jfrog/artifactory/latest/docs/resources/artifactory_user):

> Note: User passwords are never returned through the API. Since they are never returned they cannot be managed by directly through Terraform. However, it is possible to store a "known" state for the password and make changes if it's updated in Terraform. **If no password is given a random one is created otherwise that value is used. It's also worth noting "removing" the password argument does not reset the password; it just removes Terraform from storing the "known" state.**

> Note: The password is optional, but if supplied, it will be compared to the default artifactory password rules. 

and again:

> `password` - (Optional) Password for the user

**Requirements for and issue**
- [x] A fully functioning terraform snippet that can be copy&pasted (no outside files or ENV vars unless that's part of the issue)

```
resource "artifactory_user" "user" {
  count = var.create_artifactory_user ? 1 : 0

  email = var.employee_email
  name  = var.username

  groups = ["readers"]
  admin  = var.is_artifactory_admin

  # don't let people update themselves in UI; we do that from here
  profile_updatable = false
}
```

- [x] Your version of artifactory (you can `curl` it at `$host/artifactory/api/system/version`
Cloud
- [x] Your version of terraform
0.15

**Expected behavior**
I expect to be able to create users with random passwords that are not managed by TF state.  I expect a random password to be generated when I don't supply one (or to have the option to require a password reset on first use...) so that I don't need to first create a user with a random password and then in a later pass remove it.  

Alternately, update the doc to correct it.

### PR DESCRIPTION
**Pull request by danielmkn**
Fixes #260 

Readme and resource docs are modified accordingly. 
Descriptions added to all schemas elements, to make the transition to auto-generated documentation easier in the future.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -81,8 +81,10 @@ later date.
 ### Changes to user creation ###
 Previously, passwords were being generated for the user if none was supplied. This was both unnecessary (since TF has a password provider)
 and because the internal implementation could never be entirely in line with the remote server (or, be sure it was).
-With the release of 2.3.1, password is still optional, but if supplied, must watch the default password requirements. These
-can be overridden with `JFROG_PASSWD_VALIDATION_ON=false` if a custom password policy is in place.
+Then, with the release of 2.3.1, password was still optional, but if supplied, must watch the default password requirements. These
+could be overridden with `JFROG_PASSWD_VALIDATION_ON=false` if a custom password policy is in place. 
+Now this functionality is removed. Password is a required field. The verification if offloaded to the Artifactory side,
```
Comment (alexhung) [inline README.md:86]: s/if/is

Comment (alexhung) [reply README.md:86]: Also no need for `side`. Just `Artifactory` suffices.

File: docs/resources/artifactory_user.md [hunk 2]
```diff
@@ -27,14 +23,14 @@ resource "artifactory_user" "test-user" {
...
-* `admin` - (Optional) 
+* `name` - (Required) Username for user.
+* `email` - (Required) Email for user.
+* `admin` - (Optional) When enabled, this user is an administrator with all the ensuing privileges.
 * `profile_updatable` - (Optional) When set, this user can update his profile details (except for the password. Only an administrator can update the password).
 * `disable_ui_access` - (Optional) When set, this user can only access Artifactory through the REST API. This option cannot be set if the user has Admin privileges.
 * `internal_password_disabled` - (Optional) When set, disables the fallback of using an internal password when external authentication (such as LDAP) is enabled.
-* `groups` - (Optional) List of groups this user is a part of
+* `groups` - (Optional) List of groups this user is a part of.
+* `password` - (Required) Password for the user. Password validation is not done by the provider and is offloaded onto the Artifactory side. There may be cases in which you want to leave this unset to prevent users from updating their profile. For example, a departmental user with a single password shared between all department members.
```
Comment (alexhung) [inline docs/resources/artifactory_user.md:33]: Why is this moved to the bottom of this group? Same comments from the README applies here.

File: pkg/artifactory/resource_artifactory_user.go [hunk 3]
```diff
@@ -33,53 +31,57 @@ func resourceArtifactoryUser() *schema.Resource {
...
-						}
-					}
-					return nil, nil
-				},
+				Type:             schema.TypeString,
+				Sensitive:        true,
+				Required:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotEmpty),
+				Description: "Password for the user. Password validation is not done by the provider and is " +
+					"offloaded onto the Artifactory side.",
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_user.go:84]: Same comments from README applies here

**Review by alexhung — CHANGES_REQUESTED** [review]
Looks fine. Just some grammar/language changes.

File: docs/resources/artifactory_user.md (hunk 2 above)
Comment (danielmkn) [reply docs/resources/artifactory_user.md:33]: to match the order in the schema.

File: pkg/artifactory/resource_artifactory_user.go (hunk 3 above)
Comment (danielmkn) [reply pkg/artifactory/resource_artifactory_user.go:84]: fixed

File: README.md (hunk 1 above)
Comment (danielmkn) [reply README.md:86]: fixed

**Review by chb0github — APPROVED** [review]

File: docs/resources/artifactory_user.md [hunk 4]
```diff
@@ -27,14 +23,14 @@ resource "artifactory_user" "test-user" {
...
-* `admin` - (Optional) 
+* `name` - (Required) Username for user.
+* `email` - (Required) Email for user.
+* `admin` - (Optional) When enabled, this user is an administrator with all the ensuing privileges.
 * `profile_updatable` - (Optional) When set, this user can update his profile details (except for the password. Only an administrator can update the password).
 * `disable_ui_access` - (Optional) When set, this user can only access Artifactory through the REST API. This option cannot be set if the user has Admin privileges.
 * `internal_password_disabled` - (Optional) When set, disables the fallback of using an internal password when external authentication (such as LDAP) is enabled.
-* `groups` - (Optional) List of groups this user is a part of
+* `groups` - (Optional) List of groups this user is a part of.
+* `password` - (Required) Password for the user. Password validation is not done by the provider and is offloaded onto the Artifactory. There may be cases in which you want to leave this unset to prevent users from updating their profile. For example, a departmental user with a single password shared between all department members.
```
Comment (alexhung) [inline docs/resources/artifactory_user.md:33]: It feels odd to have a required attribute not grouped together with the others.

**Review by alexhung — APPROVED** [review]

**Comment by jcogilvie** [comment]
So if I want to create my users with artifactory, and make them follow the password reset procedure to actually log in, this will not try to overwrite their password after they reset it?

**Comment by danielmkn** [comment]
@jcogilvie, correct. Test scenario for your particular case:
 
1. Create a new user with Artifactory provider, set a password:

`resource "artifactory_user" "test-user" {
  name     = "terraformuser"
  email    = "email@jfrog.com"
  groups   = ["readers"]
  password = "MyPassword1"
}`

2. Set password expiration policy for your Artifactory instance, with [this call](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-SetPasswordExpirationPolicy)
Note: Without setting up an expiration policy you can't expire a user's password. 

3. Expire the password for this user, using [this API call](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-ExpirePasswordforaSingleUser)
Note: you can expire passwords for a single user, set of users, or all the users in your instance. 

4. Run `terraform apply` again. Actual result: After the first run you'll see this message `Note: Objects have changed outside of Terraform`. This is happening because this object has `Computed` fields, which are known only after the configuration was applied. It's not related to the password field. Run `terraform apply` again to double-check, and you will see the message: `No changes. Your infrastructure matches the configuration.`
Terraform won't detect (and try to undo) the password change because the password is not returned back in the API call and is not stored in the state.

Thank you for your contribution!

**Comment by jcogilvie** [comment]
My process since submitting the bug has been:  

- Make terraform generate a random password.
- Pass that into a user 
- Remove the password from the user (to remove it from state)
- Have the user run through the "forgot password" flow

So, if I rewrite my terraform at this point, with a bunch of users who have 'forgotten' passwords from state, to provide (and then forget) *another* password, this will overwrite their passwords again, yes?

Is there a way to avoid all that?  

I guess at this point it would be a feature request to generate a one-time password if a password isn't provided.

**Comment by chb0github** [comment]
@jcogilvie - Believe me, this is a big headache for us too. RT password management has no end of caveats.

As for you case: You could import state for all users, which would get you current users and their encrypted passwords. Then, you could add new users/generate passwords for them.

@danielmkn - one change that could be made would be to add a field like `on_login_pass_reset` which would force the above user reset API. You could add that as a ticket

However, @jcogilvie Daniel is right: Password policy has to be turned on globally first.

As a side note to all parties: I See that our user resource is not in alignment anymore with the UI. 

password management sucks and so far, any decision we make, someone is not happy

**Comment by jcogilvie** [comment]
> You could import state for all users, which would get you current users and their encrypted passwords. Then, you could add new users/generate passwords for them.

Unfortunately this would mean I need to maintain two categories of users indefinitely.  

Our users are modules with accounts for various services.  Everyone in the company has a user module (over a hundred).  Their (throwaway) rt passwords are `random_password` resources.  So while yes, I could import artifactory users, I *can't* import the `random_password` resources that are used to provision them, meaning that I need to indefinitely skip the `random_password` object for only *some* users.  

As of right now I can't really use this in this way.

**Comment by chb0github** [comment]
We are legit open to ideas on this subject

**Comment by jcogilvie** [comment]
Well, I think the way the documentation previously stated that it was working would have been perfectly acceptable for my use case.  

I.e., if no password is provided: 

- if the user does not exist in state, generate one, create the user, and forget the password (or even invalidate it so the user has to "forget password")
- if the user does exist in state, don't touch it

And if a password is provided:

- if the user does not exist in state, use it 
- if the user does exist in state, set the password to the new one (although I recognize that "has it changed?" is a hard question in this context; I think "I have changed my pw via terraform" could be addressed with a "force overwrite password" flag of some sort)

I don't really see much use case for storing and managing a password for a user via raw tf code because plaintext is bad, so you're pretty much going to be forced to use something generated and then plumb it in... in which case why not have your module do the generation at create time?  Maybe with a flag or something for an option to forget it from state and stop managing it in the event it changes, and/or a flag to force the password to be overwritten by what tf thinks the password should be.
